core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -289,13 +290,23 @@ private[spark] class ExecutorAllocationManager(
       s" tasksperexecutor: $tasksPerExecutor")
     val maxNeeded = math.ceil(numRunningOrPendingTasks * executorAllocationRatio /
       tasksPerExecutor).toInt
-    if (tasksPerExecutor > 1 && maxNeeded == 1 && pendingSpeculative > 0) {
+    val totalNeed = if (tasksPerExecutor > 1 && maxNeeded == 1 && pendingSpeculative > 0) {
       // If we have pending speculative tasks and only need a single executor, allocate one more
       // to satisfy the locality requirements of speculation
       maxNeeded + 1
     } else {
       maxNeeded
     }
+
+    // Request additional executors to schedule the unschedulable tasks as well
+    if (numUnschedulables > 0) {
+      val maxNeededForUnschedulables = math.ceil(numUnschedulables * executorAllocationRatio /
+        tasksPerExecutor).toInt
+      math.max(totalNeed, executorMonitor.executorCountWithResourceProfile(rpId)) +
```

Comment:
We discussed offline that using the current executor count can result in too many executors getting allocated, if the scheduling loop runs between when an executor is allocated and when tasks are scheduled on it (thus clearing out `unschedulableTaskSets`). Did you come up with any solution for this? If not and we think it's a small enough issue to ignore, let's at least add a comment about this caveat.

Reply:
Currently ExecutorAllocationManager runs every 100ms and resourceOffer happens every 1sec. `unschedulableTaskSets` gets cleared only if any task becomes schedulable and also we'll send the `unschedulable` event only once unless its removed from the `unschedulableTaskSets`. So I'm guessing it should be fine. Please correct me if I'm forgetting something.